src/WinRT.Runtime/ObjectReference.cs

```diff
@@ -1,305 +1,334 @@
+
+        public static unsafe ObjectReference<T> FromAbi(IntPtr thisPtr, T vftblT)
+        {
+            if (thisPtr == IntPtr.Zero)
+            {
+                return null;
+            }
+            var obj = new ObjectReference<T>(thisPtr, vftblT);
+            obj.VftblIUnknown.AddRef(obj.ThisPtr);
+            return obj;
+        }
+
+        public static ObjectReference<T> FromAbi(IntPtr thisPtr)
+        {
+            if (thisPtr == IntPtr.Zero)
+            {
+                return null;
+            }
+            var vftblT = GetVtable(thisPtr);
+            return FromAbi(thisPtr, vftblT);
+        }
+
+        private static unsafe T GetVtable(IntPtr thisPtr)
+        {
+            var vftblPtr = Unsafe.AsRef<VftblPtr>(thisPtr.ToPointer());
+            T vftblT;
+            // With our vtable types, the generic vtables will have System.Delegate fields
+            // and the non-generic types will have only void* fields.
+            // On .NET 5, we can use RuntimeHelpers.IsReferenceorContainsReferences
+            // to disambiguate between generic and non-generic vtables since it's a JIT-time constant.
+            // Since it is a JIT time constant, this function will be branchless on .NET 5.
+            // On .NET Standard 2.0, the IsReferenceOrContainsReferences method does not exist,
+            // so we instead fall back to typeof(T).IsGenericType, which sadly is not a JIT-time constant.
+#if NETSTANDARD2_0
+            if (typeof(T).IsGenericType)
+#else
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+#endif
+            {
+                vftblT = (T)typeof(T).GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.CreateInstance, null, new[] { typeof(IntPtr) }, null).Invoke(new object[] { thisPtr });
+            }
+            else
+            {
+                vftblT = Unsafe.AsRef<T>(vftblPtr.Vftbl.ToPointer());
+            }
+            return vftblT;
+        }
+    }
+
+    internal class ObjectReferenceWithContext<T> : ObjectReference<T>
+    {
+        private static readonly Guid IID_ICallbackWithNoReentrancyToApplicationSTA = Guid.Parse("0A299774-3E4E-FC42-1D9D-72CEE105CA57");
+        private readonly IntPtr _contextCallbackPtr;
+
+        internal ObjectReferenceWithContext(IntPtr thisPtr, IntPtr contextCallbackPtr)
+            : base(thisPtr)
+        {
+            _contextCallbackPtr = contextCallbackPtr;
+        }
+
+        protected override unsafe void Release()
+        {
+            ComCallData data = default;
+            IntPtr contextCallbackPtr = _contextCallbackPtr;
+
+            var contextCallback = new ABI.WinRT.Interop.IContextCallback(ObjectReference<ABI.WinRT.Interop.IContextCallback.Vftbl>.Attach(ref contextCallbackPtr));
+
+            contextCallback.ContextCallback(_ =>
+            {
+                base.Release();
+                return 0;
+            }, &data, IID_ICallbackWithNoReentrancyToApplicationSTA, 5);

```

Comment:
Perhaps doc what 5 is?

Reply:
it's ignored, but will add a comment.  See: https://devblogs.microsoft.com/oldnewthing/20191128-00/?p=103157